### PR TITLE
🌱Refactor controlplane health check in KCP

### PR DESCRIPTION
**Pull request by sedefsavas**
**What this PR does / why we need it**:

This PR moves reconcileControlPlaneHealth() up in the code and called in reconcile(). 

Also, it divides health check into 2 parts:
1) Only does the health check 
2) Calls the health check functions in (1) and analysis the result returned by the health check

This is partly for cleanup and partly to reuse health check in `reconcileDelete()`. With the introduction of new KCP conditions (https://github.com/kubernetes-sigs/cluster-api/pull/3674), we would like to call health checks in `reconcileDelete()` too so that we can get some signal about where the deletion process is at.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Relevant PRs: #3138 #3670 #3674

### CODE REVIEW

**Comment by sedefsavas** [comment]
cc @fabriziopandini cc @vincepri

File: controlplane/kubeadm/controllers/controller.go [hunk 1]
```diff
@@ -289,21 +288,25 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
...
 	}
 
-	ownedMachines := controlPlaneMachines.Filter(machinefilters.OwnedMachines(kcp))
-	if len(ownedMachines) != len(controlPlaneMachines) {
-		logger.Info("Not all control plane machines are owned by this KubeadmControlPlane, refusing to operate in mixed management mode")
-		return ctrl.Result{}, nil
-	}
-
-	controlPlane, err := internal.NewControlPlane(ctx, r.Client, cluster, kcp, ownedMachines)
+	controlPlane, err := r.createControlPlane(ctx, cluster, kcp)
```
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/controller.go:291]: not for this PR. 
using `controlPlane` is kind of confusing, in the future we should consider if to call this scope like in CAP

File: controlplane/kubeadm/internal/cluster.go [hunk 2]
```diff
@@ -28,22 +28,23 @@ import (
...
 type ManagementCluster interface {
 	ctrlclient.Reader
 
 	GetMachinesForCluster(ctx context.Context, cluster client.ObjectKey, filters ...machinefilters.Func) (FilterableMachineCollection, error)
-	TargetClusterEtcdIsHealthy(ctx context.Context, clusterKey client.ObjectKey) error
-	TargetClusterControlPlaneIsHealthy(ctx context.Context, clusterKey client.ObjectKey) error
+	TargetClusterControlPlaneHealthCheck(ctx context.Context, controlPlane *ControlPlane, clusterKey client.ObjectKey) (HealthCheckResult, error)
+	TargetClusterEtcdHealthCheck(ctx context.Context, controlPlane *ControlPlane, clusterKey client.ObjectKey) (HealthCheckResult, error)
+	TargetClusterAnalyseEtcdHealth(ctx context.Context, controlPlane *ControlPlane, clusterKey client.ObjectKey) error
+	TargetClusterAnalyseControlPlaneHealth(ctx context.Context, controlPlane *ControlPlane, clusterKey client.ObjectKey) error
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/cluster.go:47]: I'm not sure this fits into this PR, but I find the following points somehow confusing
- We are using the term `TargetCluster`, which is quite unusual for CAPI. Should we use `WorkloadCluster` instead?
- Why are those methods are defined on the `ManagementCluster` interface, if their goal is to access the status of elements (Etcd and ControlPlane) in a workload cluster. Should we move those methods to the `WorkloadCluster` interface instead, so we can definitely drop `Target/WorkloadCluster` the prefix?
- The difference between `HealthCheck` and `AnalyseEtcdHealth` is not really intuitive if I look at the func name only

File: controlplane/kubeadm/internal/cluster.go [hunk 3]
```diff
@@ -180,16 +181,15 @@ func (m *Management) getApiServerEtcdClientCert(ctx context.Context, clusterKey
...
+func (m *Management) healthCheck(controlPlane *ControlPlane, nodeChecks HealthCheckResult, clusterKey client.ObjectKey) error {
 	var errorList []error
-	nodeChecks, err := check(ctx)
-	if err != nil {
-		errorList = append(errorList, err)
-	}
+	kcpMachines := controlPlane.Machines.UnsortedList()
+	// Make sure Cluster API is aware of all the nodes.
+
+	// TODO: If any node has a failure, healthCheck fails. This may be too strict for the health check of HA clusters (Missing a single etcd pod does not indicate a problem)...
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/cluster.go:192]: KCP health checks are specifically designed to ensure a kubeadm generated control-plane is stable before 
proceeding with KCP actions like scale-up.

This could be seen as a strict, but given that scaling up could increase the number of members required for etcd quorum, this can lead to problems if we allow scaling up when there existing etcd members (it is the same reason behind the decision to scale in when doing remediations). 

If we are willing to change this mechanism, which is responsible for the "slowness" of KCP, I would suggest opening an issue (instead of a TODO) so we can have a proper discussion about pro and cons.

File: controlplane/kubeadm/controllers/controller.go (hunk 1 above)
Comment (vincepri) [reply controlplane/kubeadm/controllers/controller.go:291]: YES please

File: controlplane/kubeadm/controllers/controller.go [hunk 4]
```diff
@@ -31,13 +31,6 @@ import (
 	kerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/client-go/tools/record"
 	"k8s.io/utils/pointer"
-	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
-	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-	"sigs.k8s.io/controller-runtime/pkg/handler"
-	"sigs.k8s.io/controller-runtime/pkg/source"
-
```
Comment (vincepri) [inline controlplane/kubeadm/controllers/controller.go:40]: Remove this space and merge all imports together

File: controlplane/kubeadm/internal/cluster.go [hunk 5]
```diff
@@ -199,44 +199,57 @@ func (m *Management) healthCheck(ctx context.Context, check healthCheck, cluster
...
+			// The condition for this case is set by the Machine controller
 			return errors.Errorf("control plane machine %s/%s has no status.nodeRef", machine.Namespace, machine.Name)
 		}
 		if _, ok := nodeChecks[machine.Status.NodeRef.Name]; !ok {
 			return errors.Errorf("machine's (%s/%s) node (%s) was not checked", machine.Namespace, machine.Name, machine.Status.NodeRef.Name)
 		}
 	}
-	if len(nodeChecks) != len(machines) {
-		return errors.Errorf("number of nodes and machines in namespace %s did not match: %d nodes %d machines", clusterKey.Namespace, len(nodeChecks), len(machines))
+	if len(nodeChecks) != len(kcpMachines) {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/cluster.go:213]: Should we move this check up in the func, before actually running health checks?

File: controlplane/kubeadm/controllers/controller.go [hunk 6]
```diff
@@ -289,21 +288,25 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
...
-	controlPlane, err := internal.NewControlPlane(ctx, r.Client, cluster, kcp, ownedMachines)
+	controlPlane, err := r.createControlPlane(ctx, cluster, kcp)
 	if err != nil {
 		logger.Error(err, "failed to initialize control plane")
 		return ctrl.Result{}, err
 	}
+	if len(controlPlane.Machines) != len(controlPlaneMachines) {
+		logger.Info("Not all control plane machines are owned by this KubeadmControlPlane, refusing to operate in mixed management mode")
+		return ctrl.Result{}, nil
+	}
```
Comment (vincepri) [inline controlplane/kubeadm/controllers/controller.go:299]: This is a behavioral change, we should document when the scenarios when this could happen, and possibly also have this change in a different PR so it has its own release note line

File: controlplane/kubeadm/controllers/controller.go [hunk 7]
```diff
@@ -289,21 +288,25 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
...
 
 	// Aggregate the operational state of all the machines; while aggregating we are adding the
 	// source ref (reason@machine/name) so the problem can be easily tracked down to its source machine.
-	conditions.SetAggregate(controlPlane.KCP, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef())
+	conditions.SetAggregate(controlPlane.KCP, controlplanev1.MachinesReadyCondition, controlPlane.Machines.ConditionGetters(), conditions.AddSourceRef())
+
+	// reconcileControlPlaneHealth returns err if there is a machine being delete or control plane is unhealthy.
+	// If control plane is not initialized, then control-plane machines will be empty and hence health check will not fail.
+	if result, err := r.reconcileControlPlaneHealth(ctx, cluster, kcp, controlPlane); err != nil || !result.IsZero() {
```
Comment (vincepri) [inline controlplane/kubeadm/controllers/controller.go:307]:
```suggestion
	// reconcileControlPlaneHealth returns err if there is a machine being deleted or if the control plane is unhealthy.
	// If the control plane is not yet initialized, this call shouldn't fail.
	if result, err := r.reconcileControlPlaneHealth(ctx, cluster, kcp, controlPlane); err != nil || !result.IsZero() {
```

This comment might be more appropriate for the method itself, rather than here

File: controlplane/kubeadm/controllers/controller.go [hunk 8]
```diff
@@ -289,21 +288,25 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
...
-		return ctrl.Result{}, nil
-	}
-
-	controlPlane, err := internal.NewControlPlane(ctx, r.Client, cluster, kcp, ownedMachines)
+	controlPlane, err := r.createControlPlane(ctx, cluster, kcp)
 	if err != nil {
 		logger.Error(err, "failed to initialize control plane")
 		return ctrl.Result{}, err
 	}
+	if len(controlPlane.Machines) != len(controlPlaneMachines) {
```
Comment (vincepri) [inline controlplane/kubeadm/controllers/controller.go:296]: This is really confusing now to be honest

File: controlplane/kubeadm/controllers/controller.go [hunk 9]
```diff
@@ -324,7 +327,7 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	}
 
 	// If we've made it this far, we can assume that all ownedMachines are up to date
-	numMachines := len(ownedMachines)
+	numMachines := len(controlPlane.Machines)
```
Comment (vincepri) [inline controlplane/kubeadm/controllers/controller.go:330]: Make this a method on `controlPlane` struct?

File: controlplane/kubeadm/controllers/controller.go [hunk 10]
```diff
@@ -372,13 +375,52 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
...
+	}
+	ownedMachines := controlPlaneMachines.Filter(machinefilters.OwnedMachines(kcp))
+
+	controlPlane, err := internal.NewControlPlane(ctx, r.Client, cluster, kcp, ownedMachines)
+	if err != nil {
+		logger.Error(err, "failed to initialize control plane")
+		return nil, err
+	}
+	return controlPlane, nil
+}
```
Comment (vincepri) [inline controlplane/kubeadm/controllers/controller.go:395]: As mentioned in the other review, having this as a method seems out of place, it should probably be part of `internal.NewControlPlane` instead

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 11]
```diff
@@ -112,7 +120,7 @@ type HealthCheckResult map[string]error
 // controlPlaneIsHealthy does a best effort check of the control plane components the kubeadm control plane cares about.
 // The return map is a map of node names as keys to error that that node encountered.
 // All nodes will exist in the map with nil errors if there were no errors for that node.
-func (w *Workload) ControlPlaneIsHealthy(ctx context.Context) (HealthCheckResult, error) {
+func (w *Workload) ControlPlaneIsHealthy(ctx context.Context, controlPlane *ControlPlane) (HealthCheckResult, error) {
```
Comment (vincepri) [inline controlplane/kubeadm/internal/workload_cluster.go:123]: Would this better be part of the `controlPlane` struct, why is it here?

File: controlplane/kubeadm/internal/cluster.go [hunk 12]
```diff
@@ -180,16 +181,15 @@ func (m *Management) getApiServerEtcdClientCert(ctx context.Context, clusterKey
 	return tls.X509KeyPair(crtData, keyData)
 }
 
-type healthCheck func(context.Context) (HealthCheckResult, error)
-
 // HealthCheck will run a generic health check function and report any errors discovered.
 // In addition to the health check, it also ensures there is a 1;1 match between nodes and machines.
-func (m *Management) healthCheck(ctx context.Context, check healthCheck, clusterKey client.ObjectKey) error {
+// To have access to the owned control-plane machines during health checks, need to pass owningMachines here.
+func (m *Management) healthCheck(controlPlane *ControlPlane, nodeChecks HealthCheckResult, clusterKey client.ObjectKey) error {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/cluster.go:187]: I'm finding hard to follow the goal of this func, given that the description states `HealthCheck will run a generic health check function and report any errors discovered` but now the func takes in input `HealthCheckResult`, and it does not run checks anymore

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 13]
```diff
@@ -53,12 +53,20 @@ var (
...
+)
+
 // WorkloadCluster defines all behaviors necessary to upgrade kubernetes on a workload cluster
 type WorkloadCluster interface {
 	// Basic health and status checks.
 	ClusterStatus(ctx context.Context) (ClusterStatus, error)
-	ControlPlaneIsHealthy(ctx context.Context) (HealthCheckResult, error)
-	EtcdIsHealthy(ctx context.Context) (HealthCheckResult, error)
+	ControlPlaneIsHealthy(ctx context.Context, controlPlane *ControlPlane) (HealthCheckResult, error)
+	EtcdIsHealthy(ctx context.Context, controlPlane *ControlPlane) (HealthCheckResult, error)
```
Comment (vincepri) [inline controlplane/kubeadm/internal/workload_cluster.go:69]: This seems a leftover from the other PR?

File: controlplane/kubeadm/internal/cluster.go [hunk 14]
```diff
@@ -199,44 +199,57 @@ func (m *Management) healthCheck(ctx context.Context, check healthCheck, cluster
...
+	return m.healthCheck(controlPlane, checkResult, clusterKey)
+}
+
+func (m *Management) TargetClusterEtcdHealthCheck(ctx context.Context, controlPlane *ControlPlane, clusterKey client.ObjectKey) (HealthCheckResult, error) {
+	workloadCluster, err := m.GetWorkloadCluster(ctx, clusterKey)
+	if err != nil {
+		return nil, err
+	}
+	return workloadCluster.EtcdIsHealthy(ctx, controlPlane)
 }
```
Comment (vincepri) [inline controlplane/kubeadm/internal/cluster.go:245]: What are these methods doing? How are they different? Can we remove them and consolidate on `controlPlane`?

File: controlplane/kubeadm/internal/cluster.go (hunk 2 above)
Comment (vincepri) [inline controlplane/kubeadm/internal/cluster.go:47]: These seem out of place on this struct, let's revisit this

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 15]
```diff
@@ -53,12 +53,20 @@ var (
 	ErrControlPlaneMinNodes = errors.New("cluster has fewer than 2 control plane nodes; removing an etcd member is not supported")
 )
 
+// Common control-plane pod name prefixes
+const (
+	KubeAPIServerPodNamePrefix         = "kube-apiserver"
+	KubeControllerManagerPodNamePrefix = "kube-controller-manager"
+	KubeSchedulerHealthyPodNamePrefix  = "kube-scheduler"
+	EtcdPodNamePrefix                  = "etcd"
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/workload_cluster.go:61]: might be I'm wrong, but those const seems not used in this PR

File: controlplane/kubeadm/controllers/controller.go (hunk 1 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: Removed this.

File: controlplane/kubeadm/controllers/controller.go (hunk 6 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: There is no change here, we have this check already. Changed ownedMachines to controlPlane.Machines but reverted that too.

File: controlplane/kubeadm/controllers/controller.go (hunk 9 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: Reverted to use ownedMachines.

File: controlplane/kubeadm/controllers/controller.go (hunk 10 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: Removed this function as for this PR control plane is not needed in `reconcileDelete`

File: controlplane/kubeadm/internal/cluster.go (hunk 14 above)
Comment (sedefsavas) [reply controlplane/kubeadm/internal/cluster.go]: Removed all health check out of `Management` interface.

File: controlplane/kubeadm/controllers/controller.go [hunk 16]
```diff
@@ -372,13 +375,52 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	return ctrl.Result{}, nil
 }
 
+// createControlPlane gets KCP owned control plane machines and creates ControlPlane struct using it.
+func (r *KubeadmControlPlaneReconciler) createControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane) (*internal.ControlPlane, error) {
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: Removed this function as control plane is not used in `reconcileDelete` in this PR anymore.

File: controlplane/kubeadm/controllers/controller.go [hunk 17]
```diff
@@ -372,13 +375,52 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
...
 
+	controlPlane, err := r.createControlPlane(ctx, cluster, kcp)
+	if err != nil {
+		logger.Error(err, "failed to initialize control plane")
+		return ctrl.Result{}, err
+	}
+
+	// Ignore the health check results here as well as the errors, health check functions are to set health related conditions on Machines.
+	// Errors may be due to not being able to get workload cluster nodes.
+	_, err = r.managementCluster.TargetClusterControlPlaneHealthCheck(ctx, controlPlane, util.ObjectKey(cluster))
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: Removed all health check from Management interface.

File: controlplane/kubeadm/controllers/scale.go [hunk 18]
```diff
@@ -123,7 +114,8 @@ func (r *KubeadmControlPlaneReconciler) scaleDownControlPlane(
 		}
 	}
 
-	if err := r.managementCluster.TargetClusterControlPlaneIsHealthy(ctx, util.ObjectKey(cluster)); err != nil {
+	// TODO: check if this is needed after moving the health check to the main reconcile
+	if err := r.managementCluster.TargetClusterAnalyseControlPlaneHealth(ctx, controlPlane, util.ObjectKey(cluster)); err != nil {
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/scale.go]: Removed this.

File: controlplane/kubeadm/internal/cluster.go (hunk 2 above)
Comment (sedefsavas) [reply controlplane/kubeadm/internal/cluster.go]: Moved all of them out of Management interface.

File: controlplane/kubeadm/internal/cluster.go (hunk 5 above)
Comment (sedefsavas) [reply controlplane/kubeadm/internal/cluster.go]: I think surfacing individual node errors first gives more insight on what is going on then this check.

File: controlplane/kubeadm/internal/cluster.go (hunk 12 above)
Comment (sedefsavas) [reply controlplane/kubeadm/internal/cluster.go]: Changed description and moved healthcheck to `controller.go`

File: controlplane/kubeadm/controllers/controller.go [hunk 19]
```diff
@@ -372,13 +375,52 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	return ctrl.Result{}, nil
 }
 
+// createControlPlane gets KCP owned control plane machines and creates ControlPlane struct using it.
+func (r *KubeadmControlPlaneReconciler) createControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane) (*internal.ControlPlane, error) {
+	logger := r.Log.WithValues("namespace", kcp.Namespace, "kubeadmControlPlane", kcp.Name, "cluster", cluster.Name)
+
+	controlPlaneMachines, err := r.managementCluster.GetMachinesForCluster(ctx, util.ObjectKey(cluster), machinefilters.ControlPlaneMachines(cluster.Name))
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: Removed this method.

File: controlplane/kubeadm/controllers/controller.go [hunk 20]
```diff
@@ -442,21 +469,39 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
...
 // It removes any etcd members that do not have a corresponding node.
 // Also, as a final step, checks if there is any machines that is being deleted.
-func (r *KubeadmControlPlaneReconciler) reconcileHealth(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+func (r *KubeadmControlPlaneReconciler) reconcileControlPlaneHealth(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, controlPlane *internal.ControlPlane) (ctrl.Result, error) {
+	// If there is no KCP-owned control-plane machines, then control-plane has not been initialized yet.
+	if controlPlane.Machines.Len() == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// This may be a signal for etcd quorum loss too.
```
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/controller.go:481]: We can extend this comment.

File: controlplane/kubeadm/controllers/controller.go [hunk 21]
```diff
@@ -379,6 +384,28 @@ func (r *KubeadmControlPlaneReconciler) reconcileDelete(ctx context.Context, clu
 	logger := r.Log.WithValues("namespace", kcp.Namespace, "kubeadmControlPlane", kcp.Name, "cluster", cluster.Name)
 	logger.Info("Reconcile KubeadmControlPlane deletion")
 
+	// Ignore the health check results here as well as the errors, health check functions are to set health related conditions on Machines.
+	// Errors may be due to not being able to get workload cluster nodes.
+	workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		r.Log.Info("Cannot get remote client to workload cluster during delete reconciliation", "err", err.Error())
```
Comment (sedefsavas) [inline controlplane/kubeadm/controllers/controller.go:391]: This should be log level 2-3

**Comment by vincepri** [comment]
Squash?

File: controlplane/kubeadm/internal/cluster.go (hunk 3 above)
Comment (sedefsavas) [reply controlplane/kubeadm/internal/cluster.go]: Got it. Removed this todo.

File: controlplane/kubeadm/controllers/controller.go [hunk 22]
```diff
@@ -442,21 +469,39 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
...
 		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ControlPlaneUnhealthy",
 			"Waiting for control plane to pass control plane health check to continue reconciliation: %v", err)
-		return ctrl.Result{RequeueAfter: healthCheckFailedRequeueAfter}, nil
+		return ctrl.Result{}, errors.Wrap(err, "failed to pass control-plane health check")
 	}
 
 	// If KCP should manage etcd, ensure etcd is healthy.
 	if controlPlane.IsEtcdManaged() {
-		if err := r.managementCluster.TargetClusterEtcdIsHealthy(ctx, util.ObjectKey(cluster)); err != nil {
+		checkResult, err := workloadCluster.EtcdIsHealthy(ctx)
```
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/controller.go:500]: If I'm not wrong `etcdIsHealthy` won't be checked in case the `controlPlane` is not healthy, and this can lead to having etcd conditions to freeze for some times 😓 

I'm ok in keeping this for now, but this in the future we should consider how to split setting conditions (which should always happen, without block) from the actual bit of code that blocks in case there are failures. 
Ideally, we should get the second bit to rely on conditions

File: controlplane/kubeadm/controllers/controller.go [hunk 23]
```diff
@@ -482,6 +527,40 @@ func (r *KubeadmControlPlaneReconciler) reconcileHealth(ctx context.Context, clu
 	return ctrl.Result{}, nil
 }
 
+// healthCheck will analyse HealthCheckResult and report any errors discovered.
+// In addition to the health check, it also ensures there is a 1;1 match between nodes and machines.
+func healthCheck(controlPlane *internal.ControlPlane, nodeChecks internal.HealthCheckResult, clusterKey client.ObjectKey) error {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/controller.go:532]: nits, not blocking for this PR:
- clusterKey could be removed, given that we are passing the whole control place
- should this became a method of `internal.HealthCheckResult`, so this will read 
```
if err:= checkResult.Aggregate(controlPlane); err!=nil {
    ...
}
```

File: controlplane/kubeadm/controllers/fakes_test.go [hunk 24]
```diff
@@ -57,20 +55,6 @@ func (f *fakeManagementCluster) GetMachinesForCluster(c context.Context, n clien
 	return f.Machines, nil
 }
 
-func (f *fakeManagementCluster) TargetClusterControlPlaneIsHealthy(_ context.Context, _ client.ObjectKey) error {
-	if !f.ControlPlaneHealthy {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/fakes_test.go:61]: It seems we are losing the possibility of controlling the result of `ControlPlaneHealthy` func during the refactor.
If instead, this is a leftover not actually used in tests, we should remove  `ControlPlaneHealthy` field from `fakeManagementCluster`

File: controlplane/kubeadm/controllers/fakes_test.go [hunk 25]
```diff
@@ -57,20 +55,6 @@ func (f *fakeManagementCluster) GetMachinesForCluster(c context.Context, n clien
...
-		return errors.New("control plane is not healthy")
-	}
-	return nil
-}
-
-func (f *fakeManagementCluster) TargetClusterEtcdIsHealthy(_ context.Context, _ client.ObjectKey) error {
-	if !f.EtcdHealthy {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/fakes_test.go:68]: Similar comment of L61 /  `ControlPlaneHealthy`

File: controlplane/kubeadm/controllers/controller.go (hunk 22 above)
Comment (fabriziopandini) [reply controlplane/kubeadm/controllers/controller.go:500]: A possible alternative, call both the health checks and aggregate errors after

File: controlplane/kubeadm/controllers/fakes_test.go (hunk 24 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/fakes_test.go:61]: This is no longer in use as scale up test was using this before and we moved `reconcileHealth` above.
All the other tests are setting `EtcdHealthy` to true. So no need to add this if we don't plan to add more tests.

File: controlplane/kubeadm/controllers/fakes_test.go (hunk 25 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/fakes_test.go:68]: Same as above. Will do the cleanup.

File: controlplane/kubeadm/controllers/controller.go (hunk 22 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go:500]: This makes sense, will combine the errors.

File: controlplane/kubeadm/controllers/fakes_test.go (hunk 25 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/fakes_test.go:68]: I found one case that this was used to added them to fake workload cluster.

File: controlplane/kubeadm/controllers/controller.go (hunk 23 above)
Comment (sedefsavas) [reply controlplane/kubeadm/controllers/controller.go]: This makes more sense, changed.

File: controlplane/kubeadm/controllers/controller.go [hunk 26]
```diff
@@ -529,7 +532,7 @@ func (r *KubeadmControlPlaneReconciler) reconcileControlPlaneHealth(ctx context.
 
 // healthCheck will analyse HealthCheckResult and report any errors discovered.
 // In addition to the health check, it also ensures there is a 1;1 match between nodes and machines.
-func healthCheck(controlPlane *internal.ControlPlane, nodeChecks internal.HealthCheckResult, clusterKey client.ObjectKey) error {
+func healthCheck(controlPlane *internal.ControlPlane, nodeChecks internal.HealthCheckResult) error {
```
Comment (fabriziopandini) [inline controlplane/kubeadm/controllers/controller.go:535]: should this be removed and replace by `nodeChecks.Aggregate`?

**Review by fabriziopandini — COMMENTED** [review]
Only one last question, otherwise lgtm pending squash

**Comment by sedefsavas** [comment]
Squashed. Thanks for the comments @fabriziopandini @vincepri
